### PR TITLE
SI-8046 BaseTypeSeq fixes with aliases

### PR DESCRIPTION
**Pull request by retronym**
Preview/Feedback by @adriaanm and build kitteh

### CODE REVIEW

**Comment by retronym** [comment]
Still working on the first commit message. Interested in @adriaanm's opinion about these changes.

**Comment by adriaanm** [comment]
Will look at this tomorrow, but in the mean time:
`full BTS             = BTS(Three.this.Op[A],A => A,Object,Any)` --> why wasn't `s/A/Int/`?

**Comment by retronym** [comment]
> why wasn't s/A/Int/?

It can be, after ba7a072 which changes `ArgsTypeRef#transform` to be `PolyType` aware. Still feel like I'm pulling threads here. I wonder why before that change everything still worked.

**Comment by retronym** [comment]
Here's a case that still doesn't work under this patch. I'm not actually sure what if it is meant to compile.

```
trait One {
  trait Op[A]
  type Alias[A] = Op[A]
}

trait Three extends One {
  trait Op[A] extends (A => A) with super.Op[A]
  def f4[T <: Alias[Int]](f: T) = f(5)
}
```

This sort of trait overriding features in [SI-7278](https://issues.scala-lang.org/browse/SI-7278) "Unsoundness in many guises".

**Comment by soc** [comment]
@retronym Not completely related to this issue, but I'm starting to think that the most promising approach would be to enforce that a nested type in a subtype need to be a valid subtype of the nested type of the supertype. Parts of the unsoundness issues seem to be caused by parts of the compiler believing that this is the case while other parts don't check that requirement in the first place.

Would be interesting to see how many issues would go away/remain after fixing this.

**Comment by adriaanm** [comment]
With respect to overriding, traits are treated like final type members. They can't be overridden, so One#Op and Three#Op are unrelated members -- we shouldn't rebind the type alias in One to Three#Op when seen from there. The same thing holds for a type alias -- for a different reason -- as it can only be overridden by another alias with a RHS that's =:= the overridden type alias's RHS.

```
scala> trait One {
     |   trait Op[A]
     |   type Alias[A] = Op[A]
     | }
defined trait One

scala> trait Three extends One {
     |   trait Op[A] extends (A => A) with super.Op[A]
     |   // def f4[T <: Alias[Int]](f: T) = f(5)
     | }
defined trait Three

scala> val aliasSym   = typeOf[One] nonPrivateMember(newTypeName("Alias")) 
aliasSym: $r.intp.global.Symbol = type Alias

scala> val oneOpSym   = typeOf[One] nonPrivateMember(newTypeName("Op")) 
oneOpSym: $r.intp.global.Symbol = trait Op

scala> val threeOpSym = typeOf[Three] nonPrivateMember(newTypeName("Op")) 
threeOpSym: $r.intp.global.Symbol = trait Op

scala> 

scala> val threeAlias = typeOf[Three] memberType aliasSym
threeAlias: $r.intp.global.Type = [A]Three#Op[A]

scala> threeAlias.typeSymbol ne threeOpSym
res0: Boolean = true // phew!

scala> threeAlias.typeSymbol eq oneOpSym
res1: Boolean = true

scala> threeAlias.typeSymbol.ownerChain 
res2: List[$r.intp.global.Symbol] = List(trait Op, trait One, 
```

**Comment by adriaanm** [comment]
When using abstract types, the rebind correctly happens:

```
scala> trait One {
     |   type Op[A]
     |   type Alias[A] = Op[A]
     | }
scala> trait Three extends One {
     |   type Op[A] <: (A => A)
     |   // def f4[T <: Alias[Int]](f: T) = f(5)
     | }

scala> val aliasSym   = typeOf[One] nonPrivateMember(newTypeName("Alias")) 
aliasSym: $r.intp.global.Symbol = type Alias

scala> val oneOpSym   = typeOf[One] nonPrivateMember(newTypeName("Op")) 
oneOpSym: $r.intp.global.Symbol = type Op

scala> val threeOpSym = typeOf[Three] nonPrivateMember(newTypeName("Op")) 
threeOpSym: $r.intp.global.Symbol = type Op

scala> val threeAlias = typeOf[Three] memberType aliasSym
threeAlias: $r.intp.global.Type = [A]Three#Op[A]

scala> threeAlias.typeSymbol.ownerChain
res0: List[$r.intp.global.Symbol] = List(type Op, trait Three, 
```

**Comment by adriaanm** [comment]
The combination in SI-8046 (One/Three == type/trait) is also valid, of course (and Alias gets rebound to the trait in Three).

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -1999,7 +1999,9 @@ trait Types
       if (sym.typeParams.size != args.size)
         devWarning(s"$this.transform($tp), but tparams.isEmpty and args=$args")
 
-      asSeenFromOwner(tp).instantiateTypeParams(sym.typeParams, args)
+      val GenPolyType(tparams, result) = asSeenFromOwner(tp)
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala:2002]: I wonder if there's a way to make asSeenFrom kind-preserving. It looks like this code is working around the fact that a fully-applied type ref (albeit to dummy type args) gets transformed into a polytype (so now we have a type constructor)

File: src/reflect/scala/reflect/internal/Types.scala [hunk 2]
```diff
@@ -2309,7 +2311,14 @@ trait Types
       }
       thisInfo.decls
     }
-    protected[Types] def baseTypeSeqImpl: BaseTypeSeq = sym.info.baseTypeSeq map transform
+    protected[Types] def baseTypeSeqImpl: BaseTypeSeq =
+      if (sym.info.baseTypeSeq exists (_.typeSymbolDirect.isAbstractType))
+        // SI-8046 base type sequence might have more elements in a subclass, we can't map it element wise.
+        transform(sym.info).baseTypeSeq
+      else
+        // Optimization: no abstract types, we can compute the BTS of this TypeRef as an element-wise map
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala:2319]: Are there any other cases where this isn't a commutative diagram (i.e., transform then BTS != BTS then transform)?
For now, I couldn't think of any and the BTS-then-transform does indeed seem purely an optimization.

**Comment by adriaanm** [comment]
LGTM

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala:2002]: I actually feel more uneasy about this change than the other. Not that this code was much more hygienic before...

**Pull request merged**